docs/proposals-accepted/202206-block-series-deletion-compactor.md

```diff
@@ -0,0 +1,210 @@
+* If one block has too many matched series to mask, the query performance can be hugely impacted. In this case, compactor planner should be able to know the ratio of the masked series / total series. If the ratio is high, the block needs to be compacted and masked series should be removed from the block.
+
+There are other questions about how we should perform the deletion along with compaction:
+* Series are deleted at normal compaction. This might need to wait a long time and hard to apply for blocks at the max compaction level.
+* Rewrite blocks separately. This needs more coordination with compaction to avoid duplicate/unnecessary work.
+
+To overcome those challenges, we propose two options.
+
+#### Option 1: Compactor Scans Blocks
+
+Like store gateway, compactor can also scan block indexes to find the blocks matching the tombstones. Similar to the `Mask Series At Tombstone Sync Time` of store gateway, compactor can calculate the processed tombstones for each block, each tombstone and cache the results locally. As blocks and tombstones are immutable, the result needs to be calculated only once and then reused easily. As the Prometheus format tombstones are calculated, during the compaction time the compactor just needs to generate the tombstone file in the filesystem and the TSDB compactor will apply the deletion.
+
+There are two approaches about how to scan block indexes:
+1. Download all the blocks locally and check against the tombstones.
+2. Use part of the store gateway code to scan the block index and chunks on the object storage to generate `MemTombstones` result.
+
+Approach 2 should be better as only block indexes need to be downloaded, chunks data are reading from the object storage on the fly so disk space can be saved. As compactor is an offline component, it is acceptable to read from the object storage rather than downloading the whole data.
+
+The pitfall is that we need to expose part of the store gateway code like `BucketIndexReader`, `BucketChunkReader` as well as other internal code to be used by the compactor.
+
+#### Option 2: New API to Match Tombstones
+
+Since the block indexes are already downloaded by the store gateway, it can already scan the blocks and calculate the tombstones. So a way here is to expose some APIs from store gateway, and then compactor can query the APIs to know what blocks can be deleted.
+
+The gRPC API is called `Tombstone` API. The compactor can either query from the store gateway directly or query from queriers, and queriers do the API federation from multiple store gateway instances.
+
+The example gRPC format looks like below. The response is almost the same as the `MemTombstone` result.
+
+```protobuf
+service Tombstone {
+  rpc Tombstone(TombstoneRequest) returns (TombstoneResponse);
+}
+
+message TombstoneRequest {
+  int64 min_time = 1;
+  int64 max_time = 2;
+  repeated LabelMatcher matchers = 3 [(gogoproto.nullable) = false];
```

Comment:
In some way, compactor is responsible only for some blocks (some streams). Should we perhaps give that info in requests? (maybe I am overcomplicating)?

Reply:
Yeah you are totally right, I don't have any idea about this so far but if we go with the API direction we definitely need to design this.
Maybe an optional block ID array can be added to only generate tombstones for those blocks if matched? Or we can add block IDs to matchers.